### DISCUSSION TITLE
Error on generic relation

### DISCUSSION THREAD

**Discussion by haringsrob** [larastan/larastan#2094] [General]
Hey there, I have digged a bit but could not find the cause of this.

When I call something like `$model->relatedModel->someRelation()->where(function($builder) { return $builder; }`

I get the following:
```
Anonymous function should return
             Illuminate\Database\Eloquent\Relations\HasMany<App\Models\Game, App\Models\Tournament|null>
 but returns Illuminate\Database\Eloquent\Relations\HasMany<App\Models\Game, App\Models\Tournament|null>.
💡 Template type TDeclaringModel on class Illuminate\Database\Eloquent\Relations\HasMany is not covariant. Learn more: https://phpstan.org/blog/whats-up-with-template-covariantPHPStan[return.type](https://phpstan.org/error-identifiers/return.type)
```

For what I can see the types are identical? So I am unsure on how to solve this.

games() is documented with:

```
    /**
     * @return HasMany<Game, $this>
     */
```

The games has a custom query builder, maybe that is related?

```
    /**
     * @param Builder $query
     */
    public function newEloquentBuilder($query): GameQueryBuilder
    {
        return new GameQueryBuilder($query);
    }
```

The query builder itself is docblocked with:
```
/**
 * @template TModelClass of \App\Models\Game
 *
 * @extends Builder<TModelClass>
 */
class GameQueryBuilder extends Builder
```

But I have also tried with:
` * @extends \Illuminate\Database\Eloquent\Builder<\App\Models\Game>`

Any clue?

**Comment by szepeviktor** [comment]
Try adding the return type:
https://github.com/larastan/larastan/blob/9e7233d88f4f10796fadb8a2d85c5f2b55277c76/UPGRADE.md?plain=1#L207-L212

**Reply by calebdw** [reply]
If you're on Laravel 11, then look into the HasBuilder trait

**Comment by canvural** [comment]
You might be hitting [this](https://github.com/phpstan/phpstan/issues/9377) bug. Other than that I don't see any issue.
